tools/docker/generate_dockerfiles.py

```diff
@@ -851,6 +851,13 @@ def spack_env_toolchain() -> str:
 RUN spack compiler find
 RUN spack external find --all --not-buildable
 
+# Enable Spack build cache
+ARG SPACK_BUILD_CACHE=develop-2024-12-15
```

Comment:
Oh, I didn't realize that there is now a build cache for develop tags. That's awesome!
Should we maybe link it to the Spack version that we install above:

```
ARG SPACK_VERSION=develop-2024-12-15
...
git fetch --quiet --depth 1 origin ${{SPACK_VERSION}} --no-tags
...
spack mirror add ${{SPACK_VERSION}} https://binaries.spack.io/${{SPACK_VERSION}}
```

Reply:
Spack does a weekly build. I think they should be separate, otherwise we will have to wait up to a week to be able to update the Spack version.

Unfortunately, as I just mentioned below, it is not possible to use `libint` from the build cache. I tried multiple ways to force Spack to use it, but it can't satisfy all the constraints with the version provided by the build cache.

This still saves a bit of time, since several dependencies are not built but taken directly from the build cache.

Reply:
> Unfortunately, as I just mentioned below, it is not possible to use libint from the build cache. I tried multiple ways to force Spack to use it, but it can't satisfy all the constraints with the version provided by the build cache.

Presumably, we're relying too much on Ubuntu packages. With the build cache in place I'd be ready to lean in and install most packages from Spack.

> Spack does a weekly build. I think they should be separate, otherwise we will have to wait up to a week to be able to update the Spack version.

Ok, that makes sense. My main worry is that there will be some change that leads to a lot of cache misses. Maybe as a safe-guard we could time box it:

```
timeout 10m spack -e myenv concretize -f
```

In any case, all of this could also happen in a separate PR.

Reply:
> Presumably, we're relying too much on Ubuntu packages.

That is a good point. I can try to experiment locally, and see if I can come up with a combination where `libint` is indeed pulled from the cache. Another option to look into would be to setup our own [OCI registry for the build cache](https://spack.readthedocs.io/en/latest/binary_caches.html#oci-docker-v2-registries-as-build-cache).

Happy to have a look at this in a separate PR of course.

Reply:
FYI, I tried to build without tapping into Ubuntu packages but the problem persists. Spack is not very explicit in why it fails, but there must be some incompatibility between the cached `libint` dependencies and other packages. One workaround could be to use `unify: when_possible` instead of `unify: true`, but I feat this will create more problems than it solves (there will be multiple version of some libraries etc.).